diffrax/_integrate.py

```diff
@@ -123,6 +123,9 @@ def _term_compatible(
     term_structure: PyTree,
     contr_kwargs: PyTree[dict],
 ) -> bool:
+    if term_structure == LangevinTerm and isinstance(terms, LangevinTerm):
+        return True
```

Comment:
I don't think this kind of special-casing is okay. Defining custom terms like `LangevinTerm` should be a thing that users can do just as well, it shoudn't require special support in the core of Diffrax.

Reply:
Without this it throws and error. I don't know how to work around this. The issue that Langevin term is trying to solve is quite convoluted. I'll try to explain it in a way that also answers some of your later comments.

So `LangevinTerm` is not just a custom constructor for a `MultiTerm`. The issue is that any of the `AbstractLangevinSRK` solvers only need the term to provide them with the arguments `(gamma, u, grad_f)`, which uniquely determine the Underdamped Langevin Diffusion (ULD). However, all other non-Langevin solvers need to be provided with just a `MultiTerm` which defines that particular ULD. But given the vector fields provided by the `MultiTerm`, it is impossible to deduce the arguments `gamma, u, grad_f`, so we need some object which carries both these arguments and the `MultiTerm` derived from these arguments (and the fact that it is a ULD).

This means that `AbstractLangevinSRK` needs to have the term structure `LangevinTerm` (and can accept no other term), but `_term_compatible` simply fails when I do `diffeqsolve(..., ALIGN(), LangevinTerm(...), ...)`. I think the way `_term_compatible` is written right now, it is impossible for solvers to have term structures which contain custom terms. So yes, this special-casing is not okay, but `_term_compatible` as it is written now does not work as intended.

Reply:
Ah, I think I see what you need here is to actually structure things as:
```python
MultiTerm[LangevinDriftTerm, LangevinDiffusionTerm]
```
where the drift/diffusion are parameterised by `(gamma, u, grad_f)`

The Langevin-specific solvers can then be written to accept precisely `MultiTerm[LangevinDriftTerm, LangevinDiffusionTerm]` as their `term_structure`. Meanwhile we'll be still be compatible with the other general SDE solvers because we're already a `MultiTerm` of the right form.

(Actually I think technically we might need to generalise some of the SDE solvers from accepting `MultiTerm[ODETerm, AbstractTerm]` into accepting `MultiTerm[AbstractTerm[RealScalarLike], AbstractTerm]`, but that should just be a one-line change.)

Reply:
Nevermind, I found the issue:
until now `LangevinTerm.term` was not a static field, so `_term_compatible` went digging down into the PyTree instead of just treating the `LangevinTerm` as a `LangevinTerm`. All it took was to add a ` = eqx.field(static=True)`. My bad, sorry for the whole confusion.

Still, I hope it is fine if I keep the following in `_integrate.py`:
https://github.com/patrick-kidger/diffrax/blob/15c21a66206690dde470cb0d4f77b25471758d32/diffrax/_integrate.py#L1011-L1013

If not, then I can probably try doing as you suggest and change it all into a `MultiTerm[LangevinDriftTerm, LangevinDiffusionTerm]`. I feel like this is the sort of scenario where inheriting from a concrete class (i.e. `LangevinTerm` inheriting from `MultiTerm`) would make things much simpler :). Wrappers, however, need to be unwrapped, requiring changes in other parts of the codebase. (I do agree though that the pros of "concrete means final" still outweigh the cons).

Reply:
I definitely think we should avoid special-casing this one term in `_integrate.py`. The terms we ship with shouldn't really be special, and users should be able to implement their own in the same way as us. (In some sense the only two 'special' terms we have are `AbstractTerm` and `MultiTerm`.)

Do you forsee something potentially going wrong with the `MultITerm[LangevinDriftTerm, LangevinDiffusionTerm]` approach?

Reply:
Yes, there are a few things that would go wrong with `MultiTerm[LangevinDriftTerm, LangevinDiffusionTerm]`. Can we just have a call? I feel like we're spinning in circles.

Reply:
I did more digging and the thing that goes wrong with the current situation is the following:
```
diffrax/_integrate.py:170: in _check
    vf_type = eqx.filter_eval_shape(term.vf, 0.0, yi, args)
venv/lib/python3.11/site-packages/equinox/_eval_shape.py:38: in filter_eval_shape
    dynamic_out, static_out = jax.eval_shape(ft.partial(_fn, static), dynamic)
venv/lib/python3.11/site-packages/jax/_src/api.py:2808: in eval_shape
    return jit(fun).eval_shape(*args, **kwargs)
venv/lib/python3.11/site-packages/jax/_src/pjit.py:494: in eval_shape
    _, params, _, _, out_tree, _, _, _, _ = _infer_params(jit_info, args, kwargs)
venv/lib/python3.11/site-packages/jax/_src/pjit.py:622: in _infer_params
    in_shardings_flat, in_layouts_flat = _process_in_axis_resources(
venv/lib/python3.11/site-packages/jax/_src/util.py:313: in wrapper
    return cached(_IgnoreKey(), *args, **kwargs)
<string>:4: in __eq__
    ???
venv/lib/python3.11/site-packages/jax/_src/core.py:753: in __bool__
    check_bool_conversion(self)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

arr = Traced<ShapedArray(bool[2])>with<DynamicJaxprTrace(level=1/0)>

    def check_bool_conversion(arr: Array):
      if arr.size == 0:
        raise ValueError("The truth value of an empty array is ambiguous. Use"
                         " `array.size > 0` to check that an array is not empty.")
      if arr.size > 1:
>       raise ValueError("The truth value of an array with more than one element"
                         " is ambiguous. Use a.any() or a.all()")
E       ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

venv/lib/python3.11/site-packages/jax/_src/core.py:654: ValueError
```

Given how mysterious `eqx.filter_eval_shape` is to me, I have no way of debugging this further.

Reply:
I was playing around with this, and I noticed a couple things. The error only showed up if I ran more than one test. I was narrowing it down and commented everything else but one test and it worked, but 2 tests (of any kind) errored. 

My current hypothesis is that something weird is going on with the static field. I'm not 100% up to date on equinox's static field, but I believe they get cast as aux data in the module pytree and jax requires aux data to be hashable (https://stackoverflow.com/questions/78485445/why-is-custom-pytree-aux-data-traced-after-jax-jit-for-jnp-array-but-not-for). And you have a tuple with array values (the hash of which depends on the elements https://stackoverflow.com/questions/49722196/how-does-python-compute-the-hash-of-a-tuple). It seems weird that if this were true, that it wouldn't immediately error with one, but I have had weird edge case failures with hash + jax array. However, changing the args to be static, but be reliably hashable fixed it for me (see https://github.com/lockwo/diffrax/tree/langevin_test, just one commit ahead of your branch) which I why I think this is it. I don't think my solution I did there is really great, but I think this is basically an edge case of being the only term class that has array elements, so it might require some rethinking/generalization of the term class interactions (maybe as simple as adjusting some tree maps to ignore anything that isn't a specific term?). 

Also, I think I get why the Langevin term was introduced, but my question is, rather than having a MultiTerm of drift and diffusion, would making an abstract multiterm then having Langevin term inherit from that make sense? Like it is (conceptually) a multiterm, so there could be the other multiterm and the Langevinmultiterm. Just a thought I had while experimenting.

Reply:
Thanks Owen, this is really useful information! Patrick and I are having a call this Friday at 12 noon PST (8pm UK time). If you have time to join maybe @patrick-kidger could add you to the call?